redis_consumer_pending.go

```diff
@@ -0,0 +1,93 @@
+package incus
+
+import "container/list"
+import "sync"
+import "log"
+
+type RedisConsumerPendingList struct {
+	incoming <-chan RedisCommand
+	outgoing chan RedisCommand
+	pending  *list.List
+	lock     sync.Mutex
+	stats    RuntimeStats
+
+	// Used to signal to exactly one goroutine that it should wait up
+	pendingCond *sync.Cond
+}
+
+func NewRedisConsumerPendingList(consumers int, incoming <-chan RedisCommand, stats RuntimeStats, pool *redisPool) *RedisConsumerPendingList {
+	outgoing := make(chan RedisCommand)
+
+	for i := 0; i < consumers; i++ {
+		outgoingRcv := (<-chan RedisCommand)(outgoing)
+		NewRedisConsumer(outgoingRcv, pool)
+	}
+
+	pendingList := &RedisConsumerPendingList{
+		incoming:    incoming,
+		outgoing:    outgoing,
+		pending:     list.New(),
+		lock:        sync.Mutex{},
+		stats:       stats,
+		pendingCond: sync.NewCond(&sync.Mutex{}),
+	}
+
+	go pendingList.ConsumeForever()
+	go pendingList.ForwardForever()
+
+	return pendingList
+}
+
+func (r *RedisConsumerPendingList) ConsumeForever() {
+	for {
+		command := <-r.incoming
+		if DEBUG {
+			log.Println("Dequeued command in pending")
+		}
+		r.lock.Lock()
+		r.pending.PushBack(command)
+		r.lock.Unlock()
+
+		r.pendingCond.Signal()
+	}
+}
+
+func (r *RedisConsumerPendingList) ForwardForever() {
+	for {
+		r.pendingCond.L.Lock()
+
+		var pendingLength int
+		var front *list.Element = nil
+
+		for {
+			r.lock.Lock()
+			pendingLength = r.pending.Len()
+
+			if DEBUG {
+				log.Printf("There are %d pending commands", pendingLength)
+			}
+
+			r.stats.LogPendingRedisActivityCommandsListLength(pendingLength)
+
+			if pendingLength != 0 {
+				front = r.pending.Front()
+				r.pending.Remove(front)
+				r.lock.Unlock()
```

Comment:
this is a bit confusing. there are two locks (one for the `Cond` and one for the list itself). 

The `Cond` is a mechanism for waiting until the list has an element, i.e. all of the consumer goroutines except one will attempt to acquire it, and whichever succeeds will be able to dequeue from the list of pending commands. I think `cond.Signal()` unlocks the `Cond`'s lock without the caller necessarily taking it, so that it "wakes up" exactly one goroutine. 

The `r.lock` (perhaps this should be renamed to `r.listMutex`) protects against concurrent modifications by all of the consumers *and* the master goroutine that actually adds to the list.